### ISSUE TITLE
Performance regression in Starcraft Broodwar 1.16 with cnc-ddraw (with bisect)

### ISSUE THREAD

**Issue by MamiyaOtaru** [ptitSeb/box86#823]
Broodwar using the cnc-ddraw dll from https://github.com/FunkyFr3sh/cnc-ddraw/releases (release 5.0.0.0).  Opengl backend.  DLL override: WINEDLLOVERRIDES=ddraw=n,b.  

This configuration ran a lot smoother than without cnc-ddraw.  Wine without it (set to use gdi or opengl) was just not as smooth.  But now cnc-ddraw runs with a terrible framerate, since the commit below.  Looks like a complex one :-/

632a01dbe6e25cc89455df37ecb73cc2e2f2a4ea is the first bad commit
commit 632a01dbe6e25cc89455df37ecb73cc2e2f2a4ea
Author: ptitSeb <sebastien.chev@gmail.com>
Date:   Sat Feb 11 17:19:05 2023 +0100

    [DYNAREC] Use custom mutex, and improved atomic handling

 src/box86context.c            | 37 ++++++++++++++++++---------
 src/custommem.c               | 47 ++++++++++++++++++++++++-----------
 src/dynarec/arm_lock_helper.S | 51 ++++++++++++++++++++++++-------------
 src/dynarec/arm_lock_helper.h |  8 +++++-
 src/dynarec/dynablock.c       | 31 ++++++++++++-----------
 src/dynarec/dynarec_arm.c     |  6 ++++-
 src/emu/x86int3.c             | 10 ++++----
 src/emu/x86run_private.c      |  4 +--
 src/emu/x86tls.c              | 10 ++++----
 src/include/box86context.h    | 32 +++++++++++++++++++-----
 src/include/debug.h           |  2 +-
 src/include/threads.h         |  2 ++
 src/libtools/signals.c        |  2 +-
 src/libtools/threads.c        | 58 +++++++++++++++++++++++++++----------------
 src/tools/bridge.c            |  8 +++---
 15 files changed, 203 insertions(+), 105 deletions(-)

**Comment by MamiyaOtaru** [comment]
Noticed with wine 5.13 initially, tried with wine 8.01 no difference.

**Comment by MamiyaOtaru** [comment]
was launching with 
sudo --preserve-env nice -n -10 su pi -c "WINEDLLOVERRIDES=ddraw=n,b WINEPREFIX=/home/pi/wine/.wine LD_LIBRARY_PATH=/home/pi/wine/lib/wine/:/home/pi/wine/lib/ /usr/local/bin/box86 /home/pi/wine/bin/wine ./StarCraft.exe"

If I launch instead with
WINEDLLOVERRIDES=ddraw=n,b WINEPREFIX=/home/pi/wine/.wine LD_LIBRARY_PATH=/home/pi/wine/lib/wine/:/home/pi/wine/lib/ /usr/local/bin/box86 /home/pi/wine/bin/wine ./StarCraft.exe

it works fine with current Head.  sudo nice in combination with the commit above and cnc-ddraw kills performance.

Without cnc-ddraw it is fine.

Without sudo nice, it is fine

before commit https://github.com/ptitSeb/box86/commit/632a01dbe6e25cc89455df37ecb73cc2e2f2a4ea it is fine with both cnc-ddraw and sudo nice.  

Not sure if that is worth chasing down further!  cnc-ddraw is definitely good for smoother gameplay and setting the game to run with a higher priority was nice for preventing other programs from causing hitches, so it would be good to have both of them working together again, but I know this is a pretty specific trigger.

**Comment by ptitSeb** [comment]
Thanks for the analysis. I'll have a look, it shouldn't be slower.

**Comment by MamiyaOtaru** [comment]
current version, same.  Looking at it a bit closer the CPU is just being used a lot more.  Prior to that commit there was one wine Starcraft.exe process using about 50% with the rest not using much at all.  After, one wine Starcraft.exe process uses 80%+ with others using 20% and 10%.  If I use nice -10 the top one is using 90% (seems to be getting priority) which makes it worse than nice 0.  

At any rate, CPU usage is quite a bit higher after https://github.com/ptitSeb/box86/commit/632a01dbe6e25cc89455df37ecb73cc2e2f2a4ea when using cnc-ddraw.

Before:
![old](https://github.com/ptitSeb/box86/assets/2042826/6c4e8bda-2bb1-44a9-9856-7a9b8cc26684)

After:
![new](https://github.com/ptitSeb/box86/assets/2042826/dcaaf5e6-dc2a-426b-8db9-d5c1b1d68059)

After with nice -10:
![newLowNice](https://github.com/ptitSeb/box86/assets/2042826/cd032c1a-b684-4432-90b1-62eb9593f490)

**Comment by MamiyaOtaru** [comment]
Using the same cnc-ddraw with Age of Empires 2 shows a much smaller difference before that commit and after.  With nice 0 there is not a discernible difference at all.

Before (AOE2, Nice 0, cnc-ddraw)
![aoe2before](https://github.com/ptitSeb/box86/assets/2042826/d1684850-42a7-4db3-a858-c6865172f80e)

After (AOE2, Nice 0, cnc-ddraw)
![aoe2after](https://github.com/ptitSeb/box86/assets/2042826/1b4b6dc5-0d34-41cf-90ae-ad787b4822b3)

With nice -10 there is more of a difference. The busiest wine process has about the same usage (65ish), but there is a secondary wine process that uses 2 before and 8-13 after.  

Before (AOE2, Nice -10, cnc-ddraw)
![aoe2beforeLowNice](https://github.com/ptitSeb/box86/assets/2042826/729d08ba-09d3-4c33-9a26-0b70a6a8bfcf)

After (AOE2, Nice -10, cnc-ddraw)
![aoe2afterLowNice](https://github.com/ptitSeb/box86/assets/2042826/5bf3159e-9300-4b3c-b66f-2fb8cdedf52a)

Not nearly as big a difference as with Starcraft, and not noticeably detrimental in game with or without nice -10 (I like -10 is there are occasional hitches without it)

tl;dr Starcraft + cnc-ddraw is worse now.  AOE2 is the same.  Nice -10 makes Starcraft even worse now.  It also makes AOE2 worse now, but not noticeably in game.

**Comment by ptitSeb** [comment]
I thought the custom mutex were optionnal in box86, but there are not. I'll make them optionnal (at compile time) on box86 (they are in box64), so you can build a version with regular mutex, so you can compare both version.
It's still strange that there is a big speed impact, as thoses mutex are using for internal stuff that are not supposed to wait for long period of time.

**Comment by ptitSeb** [comment]
So, I have disable the use of custom mutexes for now @MamiyaOtaru so, you can update box86 and try again. I assume you should not see the large decrease in performances now.
